Fix older Mac build that still use makedepend [1.0.2]

tmshort opened the pull request:
Fixes an issue with PR #4755.

Xcode on Mac OS X 10.7 and earlier still uses `makedepend`, its C compiler does not support the `-M` option.
This was fixed with Xcode on Mac OS X 10.8, which added the capability to do `cc -M`.
<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

dot-asm commented on Configure:
```diff
@@ -1670,6 +1670,7 @@ while (<PIPE>) {
   $predefined{$1} = defined($2) ? $2 : "";
 }
 close(PIPE);
+my $old_xcode_uses_makedepend = defined($predefined{__APPLE_CC__}) && !defined($predefined{__clang__});
```
As first reaction I wanted to suggest to assign more generic name to this variable, like no_cc_as_makedepend, but then I figured that it would probably be more appropriate to consolidate the decision logic. Something like this perhaps:
```
my $makedepprog;
if ($predefined{__GNUC__} >= 3
   && !($predefined{__APPLE_CC__}) && !$predefined{__clang__})) {
    $makedepprog = $cc;
}
...
    s/^MAKEDEPPROG=.*$/MAKEDEPPROG = $makedepprog/ if defined($makedepprog);
```
One can also resolve that cross-compile prefix thing in the first if...

tmshort replied:
Yeah, I wasn't sure if I should make it a bit more generic, or keep it specific to OS X.

With respect to the cross-compile prefix, do you mean that `$randlib` and `$ar`, etc are missing in the variable substitutions?

I wasn't planning on fixing the cross-compile section, as it might actually have a cc that supports -M...

dot-asm replied:
I didn't mean that you would revamp *that* if, only make cross-compile assignment to MAKEDEPEND in *this* if. Which would allow to move MAKEDEPPROG substitution *alone* out *that* if. But this is optional! If in doubt, just make the variable more generic and consolidate logic. Just in case, rationale behind assigning generic name is "what if there is another special case like this".

tmshort replied:
The cross-compile section still looks wrong, however; it's missing the `$ar` and `$randlib`, etc variables. But, that's out of scope here.

tmshort commented:
Not sure if something similar is required for 1.1.0...

dot-asm commented on Configure:
```diff
@@ -1671,6 +1671,13 @@ while (<PIPE>) {
+# Xcode did not handle $cc -M before clang support
+my $makedepprog;
+if ($predefined{__GNUC__} >= 3 && !(defined($predefined{__APPLE_CC__})
+                                    && !defined($predefined{__clang__}))) {
```
Well, if you play `defined($predefined{})` game, then it makes more sense to check for `defined($predefined{__GNUCC__})`. Because undefined $foo in *logical* expression implies check for defined. I mean `use warnings; my $a; 1 if ($a)` triggers no warnings. Of course in `__GNUC__` case it's sufficient to `($predefined{__GNUC__} // -1)` instead of dragging whole defined().

tmshort replied:
When checking the value of `$predefined{__GNUC__}`, the rest of the Configure script did not include `defined()`. I'm just following existing usage here. :)

dot-asm replied:
Do note that I said "*if* you play the game" :-)

dot-asm commented on Configure:
```diff
@@ -1737,7 +1744,7 @@ while (<IN>)
 		s/^RANLIB=.*/RANLIB= $ranlib/;
 		s/^RC=.*/RC= $windres/;
-		s/^MAKEDEPPROG=.*$/MAKEDEPPROG= $cc/ if $predefined{__GNUC__} >= 3;
+		s/^MAKEDEPPROG=.*$/MAKEDEPPROG= $makedepprog/ if defined($makedepprog);
```
When made remark about "this" and "that" ifs, I mean that you'd replace `$predefined{__GNUC__} >= 3` with `if $makedepprog` in *both* cases. This is because we collect pre-defines from $cross_compile_prefix$cc, not just $cc. Note that workaround is likely to be needed even with `__APPLE_CC__`, because that's what's used even for ios targets, and you would set cross-compile prefix...

tmshort commented:
Rebased... I changed how the `$cross_compile_prefix` is handled; I'm not 100% sure it's the right thing to do, but it seems to make sense to me, and I believe it may be something @dot-asm has been alluding to.

dot-asm commented on Configure:
```diff
@@ -1723,21 +1730,19 @@ while (<IN>)
-		s/^RANLIB=.*/RANLIB= $ranlib/;
-		s/^RC=.*/RC= $windres/;
-		s/^MAKEDEPPROG=.*$/MAKEDEPPROG= $cc/ if $predefined{__GNUC__} >= 3;
+		s/^MAKEDEPPROG=\s*/MAKEDEPPROG= \$\(CROSS_COMPILE\)/;
```
Original possible outcomes were makedepend or $(CROSS_COMPILE)$cc. Now $(CROSS_COMPILE)makedepend becomes a possibility, and I don't see that it's viable possibility. Then originally the only thing that could happen when cross-compile prefix is defined was that it's injected into AR=ar, NM=nm, RANLIB=ranlib, RC=windres. In other words you had no option to replace ar, nm, ranlib and windres. Now it becomes possibility. One can argue that it's meaningful thing to do, but one can as well argue that long-term support branch is not right place for that kind of modifications. Or in other words modifications should preserve original behaviour.

tmshort replied:
I hear you; I thought it went a bit far, too, but decided to try it. As you point out, might very well be "correct" behavior, but too much for LTS.

tmshort commented:
Rebased again, going back to the other extreme, only modifying the MAKEDEPPROG variable in that one if, leaving it alone in the cross-compile case.

dot-asm commented:
> leaving it alone in the cross-compile case.

That would still be wrong. Indeed, MAKEDEPENDPROG is reassigned in both cases. As currently suggested it will be reassigned in one case based on availability of $makedepprog variable, and in another - on `$predefined{__GNU__} >= 3`. Now, what was the reason for suggesting a generic name for $makedepprog instead of originally suggested osx-centric? It was "what if there is another special case like this" [which would be handled by extending condition for assigning value to $makedepprog]. Since %predefined is filled with output from **$cross_compile_prefix**$cc, possibility that another case could be cross-compiled is one to consider. This in turn means that same condition, based on availability of $makedepprog, should apply in both cases, not two different. Moreover! iphone/ios targets were mentioned. Those are cross-compiled and use `__APPLE_CC__` compiler. And since there is no evidence that it would be different from "regular" compiler [in respect to quirky limitations like one in question], it's only appropriate to use same condition in both cases even in current case.

I'd like to put it this way. If you won't get it right in next commit, I'm going to take over this...

dot-asm commented:
Yes to all three questions. [Well, modulo fact that defined checks are excessive in 0. And modulo comment below.]

> (as you believe there's no cross-compile version of makedepend)

I have *never* seen one. Cross-compile toolchains customarily consist of compilers, binutils, headers and libraries. BTW, recall that makedepend is originally X11 thing...

> and the value of $(CROSS_COMPILE)$cc (since $makedepprog = $cc) when it is defined.

Well, one can assign $(CROSS_COMPILE)$makedepprog as well, because $cc is the only possible value. This is by the way why I at some point suggested to take this assignment out of "that" if. I mean you'd assign $makedepprog to either $cc or $(CROSS_COMPILE)$cc in *first* if, and then just replace MAKEDEPPROG = $makedepprog if defined($makedepprog) [taking it out second if]. This way it's most readable as it asks for least amount of questions.

tmshort commented:
Gotcha. 

I was also conflating your terms "this if" and "that if", with the "if" case and "else" case; which cause some of this churn.

tmshort commented:
I'm torn between having one modification of MAKEDEPPROG (as suggested) and keeping the structure the way it is; with the cross-compile code in one place. Both have their benefits...

dot-asm commented:
As long as variable is called $makedepprog single MAKEDEPPROG replacement is most suitable. If you want to keep two replacements, then call variable for $cc_as_makedepend and replace just `if ($predefined{__GNUC__} >= 3)` with `if ($cc_as_makedepend)` and keep replacement themselves.

dot-asm commented:
Editorial note. I'd like to make commit message less specific. Something like "Earlier Apple Xcode compilers, e.g. one targeting Mac OS X 10.7, don't support dependency generation and one still has to use makedepend. It's unclear when it was fixed, but all clang-based Apple compilers seem to support -M options." Basically objection is that Xcode version is not actually bound to specific Mac OS X, so you can't really say "Xcode **on** MacOS X Y.Z", you do have certain freedom to choose Xcode version...

dot-asm commented:
You don't have to do anything, I can change the commit message upon actual commit.

There also was question about 1.1.x. Well, it boils down to question what are users' expectations. If 1.1.x is expected to be compiled with such old compiler, then it would be appropriate to address this even in 1.1.x. The makedepend detection is different in 1.1.0 and master, so *if* it will be addressed, it would have to be separate requests... I for one would say it would be appropriate...

tmshort commented:
Yes, I didn't think this could be directly ported to 1.1.x

tmshort commented:
Waiting on a second review? @richsalz

dot-asm commented:
Merged. Thanks!
